src/plot_api/subroutines.js

```diff
@@ -397,24 +399,156 @@ function findCounterAxisLineWidth(ax, side, counterAx, axList) {
+    var x = getMainTitleX(fullLayout, textAnchor);
 
+    // Draw title without positioning to get size
     Titles.draw(gd, 'gtitle', {
         propContainer: fullLayout,
         propName: 'title.text',
         placeholder: fullLayout._dfltTitle.plot,
-        attributes: {
-            x: getMainTitleX(fullLayout, textAnchor),
-            y: getMainTitleY(fullLayout, dy),
+        attributes: ({
+            x: x,
+            y: y,
             'text-anchor': textAnchor,
             dy: dy
-        }
+        })
     });
+
+    if(title.text && title.automargin) {
+        var titleObj = d3.selectAll('.gtitle');
+        var titleHeight = Drawing.bBox(titleObj.node()).height;
+        var pushMargin = needsMarginPush(gd, title, titleHeight);
+        if(pushMargin > 0) {
+            setDflts(title, getDflts(title)[0], getDflts(title)[1]);
+            // Recalculate these since the defaults have changed
+            dy = getMainTitleDy(fullLayout);
+            y = getMainTitleY(fullLayout, dy);
+            applyTitleAutoMargin(gd, y, pushMargin, titleHeight);
+
+            // Position the title once we know where it needs to be
+            titleObj.attr({
+                x: x,
+                y: y,
+                'text-anchor': textAnchor,
+                dy: dy
+            }).call(svgTextUtils.positionText, x, y);
+        }
+    }
 };
 
+
+function isOutsideContainer(gd, title, position, y, titleHeight) {
+    var plotHeight = title.yref === 'paper' ? gd._fullLayout._size.h : gd._fullLayout.height;
+    var yPosTop = Lib.isTopAnchor(title) ? y : y - titleHeight; // Standardize to the top of the title
+    var yPosRel = position === 'b' ? plotHeight - yPosTop : yPosTop; // Position relative to the top or bottom of plot
+    if((Lib.isTopAnchor(title) && position === 't') || Lib.isBottomAnchor(title) && position === 'b') {
+        return false;
+    } else {
+        return yPosRel < titleHeight;
+    }
+}
+
+
+// title.y is 1 or 0 if automargin and paper ref
+// 'auto' is not supported for either title.y or title.yanchor when automargin=true
+function getDflts(title) {
+    var titleY = title.y;
+    var titleYanchor = title.yanchor;
+    if(title.automargin && title.yref === 'paper') {
+        titleY = title.y === 0 ? 0 : 1;
+        if(title.yanchor === 'auto') {
+            titleYanchor = title.y === 0 ? 'top' : 'bottom';
+        }
+    }
+    if(title.automargin && title.yref === 'container') {
+        if(title.y === 'auto') titleY = 1;
+        if(title.yanchor === 'auto') {
+            titleYanchor = title.y < 0.5 ? 'bottom' : 'top';
+        }
+    }
+    return [titleY, titleYanchor];
+}
+
+function setDflts(title, titleY, titleYanchor) {
+    if(title.yref === 'paper' && title.y !== 0 && title.y !== 1 && title.y !== 'auto') {
+        Loggers.warn('title.automargin=true so resetting the supplied title.y value to 1.');
+    }
+    title.y = titleY;
+    title.yanchor = titleYanchor;
+}
```

Comment:
@hannahker wondering if you may possibly consider picking this commit: https://github.com/plotly/plotly.js/commit/0a23e1d19bab780aad6a1941d5d281cb31c0fe4c
which moves this logic to the supply defaults step?

Reply:
Thanks @archmoj! [This piece](https://github.com/plotly/plotly.js/pull/6428/files#diff-d670f85239a0f2240b21379a074593658aa18b142d90eeeab8a10699713eb148R426) is what's making me think that we can't move the default setting to the supply defaults step, as we'll only do it if the margin actually needs to be pushed (addressing the scenario in your screenshots above)